File: lib/umya_spreadsheet/image_functions.ex
defmodule UmyaSpreadsheet.ImageFunctions do
  @moduledoc """
  Functions for working with images in a spreadsheet.
  """

  alias UmyaSpreadsheet.Spreadsheet
  alias UmyaSpreadsheet.ErrorHandling
  alias UmyaNative

  @doc """
  Adds an image to a spreadsheet.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet
  - `cell_address` - The cell address (e.g., "A1") where the top-left corner of the image should be placed
  - `image_path` - The path to the image file

  ## Returns

  - `:ok` on success
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      :ok = UmyaSpreadsheet.ImageFunctions.add_image(spreadsheet, "Sheet1", "C5", "path/to/image.png")
  """
  def add_image(%Spreadsheet{reference: ref}, sheet_name, cell_address, image_path) do
    UmyaNative.add_image(ref, sheet_name, cell_address, image_path)
    |> ErrorHandling.standardize_result()
  end

  @doc """
  Extracts an image from a spreadsheet and saves it to a file.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet
  - `cell_address` - The cell address (e.g., "A1") where the image is located
  - `output_path` - The path where the image should be saved

  ## Returns

  - `:ok` on success
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      :ok = UmyaSpreadsheet.ImageFunctions.download_image(spreadsheet, "Sheet1", "C5", "extracted_image.png")
  """
  def download_image(%Spreadsheet{reference: ref}, sheet_name, cell_address, output_path) do
    UmyaNative.download_image(ref, sheet_name, cell_address, output_path)
    |> ErrorHandling.standardize_result()
  end

  @doc """
  Changes an existing image in a spreadsheet to a new image.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet
  - `cell_address` - The cell address (e.g., "A1") where the image is located
  - `new_image_path` - The path to the new image file

  ## Returns

  - `:ok` on success
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      :ok = UmyaSpreadsheet.ImageFunctions.change_image(spreadsheet, "Sheet1", "C5", "path/to/new_image.png")
  """
  def change_image(%Spreadsheet{reference: ref}, sheet_name, cell_address, new_image_path) do
    UmyaNative.change_image(ref, sheet_name, cell_address, new_image_path)
    |> ErrorHandling.standardize_result()
  end

  @doc """
  Gets the dimensions (width, height) of an image in a cell.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet
  - `cell_address` - The cell address (e.g., "A1") where the image is located

  ## Returns

  - `{:ok, {width, height}}` where width and height are in pixels
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      {:ok, {width, height}} = UmyaSpreadsheet.ImageFunctions.get_image_dimensions(spreadsheet, "Sheet1", "C5")
  """
  def get_image_dimensions(%Spreadsheet{reference: ref}, sheet_name, cell_address) do
    UmyaNative.get_image_dimensions(ref, sheet_name, cell_address)
    |> ErrorHandling.standardize_result()
  end

  @doc """
  Lists all images in a sheet with their positions and names.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet

  ## Returns

  - `{:ok, [{coordinate, image_name}, ...]}` with a list of image coordinates and names
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      {:ok, images} = UmyaSpreadsheet.ImageFunctions.list_images(spreadsheet, "Sheet1")

      # Print all images in the sheet
      Enum.each(images, fn {coord, img_name} ->
        IO.puts("Image '\#{img_name}' at position \#{coord}")
      end)
  """
  def list_images(%Spreadsheet{reference: ref}, sheet_name) do
    UmyaNative.list_images(ref, sheet_name)
    |> ErrorHandling.standardize_result()
  end

  @doc """
  Gets comprehensive information about an image in a spreadsheet.

  ## Parameters

  - `spreadsheet` - The spreadsheet struct
  - `sheet_name` - The name of the sheet
  - `cell_address` - The cell address (e.g., "A1") where the image is located

  ## Returns

  - `{:ok, {name, position, width, height}}` with the image details
  - `{:error, reason}` on failure

  ## Examples

      {:ok, spreadsheet} = UmyaSpreadsheet.read_file("input.xlsx")
      {:ok, {name, pos, width, height}} = UmyaSpreadsheet.ImageFunctions.get_image_info(spreadsheet, "Sheet1", "C5")

      # Print the image information
      IO.puts("Image '" <> name <>"' at position " <> pos <>", dimensions: "<>width{}"x"<>height<>" pixels")
  """
  def get_image_info(%Spreadsheet{reference: ref}, sheet_name, cell_address) do
    UmyaNative.get_image_info(ref, sheet_name, cell_address)
    |> ErrorHandling.standardize_result()
  end
end

File: native/umya_native/src/image_functions.rs
use rustler::{Atom, Error as NifError, NifResult, ResourceArc};
use std::path::Path;
use umya_spreadsheet::structs::drawing::spreadsheet::MarkerType;
use umya_spreadsheet::structs::Image;

use crate::atoms;
use crate::UmyaSpreadsheet;

/// Helper function to ensure a worksheet is deserialized
fn ensure_worksheet_deserialized(
    guard: &mut std::sync::MutexGuard<umya_spreadsheet::Spreadsheet>,
    sheet_index: &usize,
) {
    // Call read_sheet to deserialize the worksheet
    // This is safe to call even if already deserialized
    guard.read_sheet(*sheet_index);
}

/// Add an image to a sheet
#[rustler::nif]
pub fn add_image(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
    cell_address: String,
    image_path: String,
) -> NifResult<Atom> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    match guard.get_sheet_by_name_mut(&sheet_name) {
        Some(sheet) => {
            let path = Path::new(&image_path);
            if !path.exists() {
                return Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                ))));
            }

            // Create a new image
            let mut image = Image::default();

            // Create marker and set coordinate
            let mut marker = MarkerType::default();
            marker.set_coordinate(&cell_address);

            // Add the image to the Image struct using umya-spreadsheet API
            image.new_image(&image_path, marker);

            // Add the image to the worksheet
            sheet.add_image(image);
            Ok(atoms::ok())
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}

/// Download an image from a sheet
#[rustler::nif]
pub fn download_image(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
    cell_address: String,
    output_path: String,
) -> NifResult<Atom> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    match guard.get_sheet_by_name_mut(&sheet_name) {
        Some(sheet) => {
            // Check if there's an image at this location using umya-spreadsheet API
            match sheet.get_image(cell_address.as_str()) {
                Some(image) => {
                    // Check if the image actually has data
                    if !image.has_image() {
                        return Err(NifError::Term(Box::new((
                            atoms::error(),
                            "Image not found".to_string(),
                        ))));
                    }

                    // Create the directory if it doesn't exist
                    if let Some(parent) = Path::new(&output_path).parent() {
                        if let Err(_) = std::fs::create_dir_all(parent) {
                            return Err(NifError::Term(Box::new((
                                atoms::error(),
                                "Failed to create output directory".to_string(),
                            ))));
                        }
                    }

                    // Use umya-spreadsheet's download_image method
                    image.download_image(&output_path);
                    Ok(atoms::ok())
                }
                None => Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                )))),
            }
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}

/// Change an image in a cell
#[rustler::nif]
pub fn change_image(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
    cell_address: String,
    new_image_path: String,
) -> NifResult<Atom> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    match guard.get_sheet_by_name_mut(&sheet_name) {
        Some(sheet) => {
            // Check if the new image file exists
            let path = Path::new(&new_image_path);
            if !path.exists() {
                return Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                ))));
            }

            // Check if there's an image at this location
            match sheet.get_image_mut(cell_address.as_str()) {
                Some(image) => {
                    // Check if the image actually has data
                    if !image.has_image() {
                        return Err(NifError::Term(Box::new((
                            atoms::error(),
                            "Image not found".to_string(),
                        ))));
                    }

                    // Use umya-spreadsheet's change_image method
                    image.change_image(&new_image_path);
                    Ok(atoms::ok())
                }
                None => Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                )))),
            }
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}

/// Get dimensions (width, height) of an image in a cell
#[rustler::nif]
pub fn get_image_dimensions(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
    cell_address: String,
) -> NifResult<(u32, u32)> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    // Find sheet index by name
    let sheet_index = match guard
        .get_sheet_collection_no_check()
        .iter()
        .position(|sheet| sheet.get_name() == sheet_name)
    {
        Some(index) => index,
        None => {
            return Err(NifError::Term(Box::new((
                atoms::error(),
                "Sheet not found".to_string(),
            ))))
        }
    };

    // Ensure the worksheet is deserialized before accessing it
    ensure_worksheet_deserialized(&mut guard, &sheet_index);

    match guard.get_sheet_by_name(&sheet_name) {
        Some(sheet) => {
            match sheet.get_image(cell_address.as_str()) {
                Some(image) => {
                    // Extract width and height from the image
                    let mut width = 0;
                    let mut height = 0;

                    // Width and height are stored in the extent of the anchor
                    if let Some(anchor) = image.get_one_cell_anchor() {
                        // Convert from EMU (English Metric Units) to pixels
                        // 1 pixel = 9525 EMU
                        width = (anchor.get_extent().get_cx() / 9525) as u32;
                        height = (anchor.get_extent().get_cy() / 9525) as u32;
                    } else if let Some(_anchor) = image.get_two_cell_anchor() {
                        // For two cell anchors, calculate from the from/to markers
                        let from_col = *image.get_from_marker_type().get_col();
                        let from_row = *image.get_from_marker_type().get_row();

                        if let Some(to_marker) = image.get_to_marker_type() {
                            let to_col = *to_marker.get_col();
                            let to_row = *to_marker.get_row();

                            // Simple difference estimation (this is approximate)
                            width = (to_col - from_col) * 64; // Approximate cell width
                            height = (to_row - from_row) * 20; // Approximate cell height
                        }
                    }

                    Ok((width, height))
                }
                None => Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                )))),
            }
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}

/// List all images in a sheet with their coordinates
#[rustler::nif]
pub fn list_images(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
) -> NifResult<Vec<(String, String)>> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    // Find sheet index by name
    let sheet_index = match guard
        .get_sheet_collection_no_check()
        .iter()
        .position(|sheet| sheet.get_name() == sheet_name)
    {
        Some(index) => index,
        None => {
            return Err(NifError::Term(Box::new((
                atoms::error(),
                "Sheet not found".to_string(),
            ))))
        }
    };

    // Ensure the worksheet is deserialized before accessing it
    ensure_worksheet_deserialized(&mut guard, &sheet_index);

    match guard.get_sheet_by_name(&sheet_name) {
        Some(sheet) => {
            let image_collection = sheet.get_image_collection();
            let mut result = Vec::new();

            for image in image_collection {
                // Get coordinate (cell reference like "A1")
                let coordinate = image.get_coordinate();

                // Get image name
                let name = image.get_image_name().to_string();

                result.push((coordinate, name));
            }

            Ok(result)
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}

/// Get comprehensive information about an image
#[rustler::nif]
pub fn get_image_info(
    resource: ResourceArc<UmyaSpreadsheet>,
    sheet_name: String,
    cell_address: String,
) -> NifResult<(String, String, u32, u32)> {
    let mut guard = resource.spreadsheet.lock().unwrap();

    // Find sheet index by name
    let sheet_index = match guard
        .get_sheet_collection_no_check()
        .iter()
        .position(|sheet| sheet.get_name() == sheet_name)
    {
        Some(index) => index,
        None => {
            return Err(NifError::Term(Box::new((
                atoms::error(),
                "Sheet not found".to_string(),
            ))))
        }
    };

    // Ensure the worksheet is deserialized before accessing it
    ensure_worksheet_deserialized(&mut guard, &sheet_index);

    match guard.get_sheet_by_name(&sheet_name) {
        Some(sheet) => {
            match sheet.get_image(cell_address.as_str()) {
                Some(image) => {
                    // Get basic info
                    let name = image.get_image_name().to_string();
                    let position = image.get_coordinate();

                    // Get dimensions
                    let mut width = 0;
                    let mut height = 0;

                    // Width and height are stored in the extent of the anchor
                    if let Some(anchor) = image.get_one_cell_anchor() {
                        // Convert from EMU (English Metric Units) to pixels
                        // 1 pixel = 9525 EMU
                        width = (anchor.get_extent().get_cx() / 9525) as u32;
                        height = (anchor.get_extent().get_cy() / 9525) as u32;
                    } else if let Some(_anchor) = image.get_two_cell_anchor() {
                        // For two cell anchors, calculate from the from/to markers
                        let from_col = *image.get_from_marker_type().get_col();
                        let from_row = *image.get_from_marker_type().get_row();

                        if let Some(to_marker) = image.get_to_marker_type() {
                            let to_col = *to_marker.get_col();
                            let to_row = *to_marker.get_row();

                            // Simple difference estimation (this is approximate)
                            width = (to_col - from_col) * 64; // Approximate cell width
                            height = (to_row - from_row) * 20; // Approximate cell height
                        }
                    }

                    Ok((name, position, width, height))
                }
                None => Err(NifError::Term(Box::new((
                    atoms::error(),
                    "Image not found".to_string(),
                )))),
            }
        }
        None => Err(NifError::Term(Box::new((
            atoms::error(),
            "Sheet not found".to_string(),
        )))),
    }
}
